File: usecase-2.1/work/pipeline.m
function pipeline(varargin)

% This pipeline computes averages from the participants.tsv file
%
% Use as 
%    pipeline [options] <inputdir> <outputdir> <level>
% where the input and output directory must be specified, and the 
% level is either "group" or "participant".
%
% Optional arguments:
%   -h,--help           Show this help and exit.
%   --verbose           Enable verbose output.
%   --start-idx <num>   Start index for participant selection.
%   --stop-idx <num>    Stop index for participant selection.

% This code is shared under the CC0 license
%
% Copyright (C) 2024, SIESTA workpackage 15 team

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% parse the command-line options
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% optional arguments can be "flags" that come by themselves, like '-h' or '--help'
% optional arguments can also have values, like '--species {human,rat}'

% flags are represented as true/false
options.help       = false;
options.verbose    = false;
% other options each have their own value, either a string or a number
options.start_idx  = [];
options.stop_idx   = [];

% deal with the flags
isflag = false(size(varargin));
for i=1:numel(varargin)
  switch varargin{i}
    case {'-h', '--help'}
      options.help = true;
      isflag(i) = true;
    case {'-v', '--verbose'}
      options.verbose = true;
      isflag(i) = true;
  end % switch
end % for

% remove the flags
varargin = varargin(~isflag);
clear isflag

% deal with the optional arguments
isoption = false(size(varargin));
for i=1:2:numel(varargin)
  switch varargin{i}
    case {'--start-idx'}
      options.start_idx = str2double(varargin{i+1});
      isoption(i) = true;
      isoption(i+1) = true;
    case {'--stop-idx'}
      options.stop_idx = str2double(varargin{i+1});
      isoption(i) = true;
      isoption(i+1) = true;
  end % switch
end % for

% remove the optional arguments
varargin = varargin(~isoption);
clear isoption

% show the help (if requested)
if options.help
  help(mfilename);
  return
end

% deal with the positional arguments
if length(varargin)<3
  error('not enough input arguments')
elseif length(varargin)>3
  str = sprintf('%s ', varargin{1:end-3});
  error('unsupported input arguments: %s ', str);
end

% the last three arguments must be inputdir, outputdir and level
inputdir  = varargin{end-2};
outputdir = varargin{end-1};
level     = varargin{end}; % participant or group

% the last three arguments must be strings
if ~isa(inputdir, 'char') && ~isa(inputdir, 'string')
  error('incorrect specification of inputdir');
elseif ~isa(outputdir, 'char') && ~isa(outputdir, 'string')
  error('incorrect specification of inputdir');
elseif ~isa(level, 'char') && ~isa(level, 'string')
  error('incorrect specification of inputdir');
end

if ~strcmpi(level, 'participant') && ~strcmpi(level, 'group')
  error('level should either be ''participant'' or ''group''');
end

if ~exist(inputdir, 'dir')
  error('input directory does not exist');
end

if ~exist(outputdir, 'dir')
  warning('creating output directory');
  [success, message] = mkdir(outputdir);
  if ~success
    error(message);
  end
end

% add these to the options structure
options.inputdir  = inputdir;
options.outputdir = outputdir;
options.level     = level;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% call the actual code to execute the pipeline
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

if options.verbose
  fprintf('options =\n');
  disp(options);
end

if strcmp(options.level, 'participant')
  disp("nothing to do at the participant level")
  return
end

inputfile  = fullfile(options.inputdir, 'participants.tsv');
outputfile = fullfile(options.outputdir, 'results.tsv');

participants = readtable(inputfile, 'FileType', 'text', 'Delimiter', '\t', 'VariableNamingRule', 'preserve');

if options.verbose
  fprintf('data contains %d participants\n', size(participants, 1));
end

% select participants
if ~isempty(options.stop_idx)
  participants = participants(1:options.stop_idx,:);
end
if ~isempty(options.start_idx)
  participants = participants(options.start_idx:end,:);
end

if options.verbose
  fprintf('selected %d participants\n', size(participants, 1));
end

averagedage  = mean(participants.age, 'omitnan');
averagedHeight = mean(participants.Height, 'omitnan');
averagedWeight = mean(participants.Weight, 'omitnan');

% put the results in a table
result = table(averagedage, averagedHeight, averagedWeight);

if options.verbose
  disp(result);
end

if options.verbose
  disp(['writing to ' outputfile]);
end

writetable(result, outputfile, 'FileType', 'text', 'Delimiter', '\t', 'WriteVariableNames', false);

File: usecase-2.1/work/pipeline.R
#!/usr/bin/env Rscript

# This pipeline computes averages from the participants.tsv file
#
# Use as
#    ./pipeline.R [options] <inputdir> <outputdir> <level>
# where the input and output directory must be specified, and the
# level is either "group" or "participant".
#
# Optional arguments:
#   -h,--help           Show this help and exit.
#   --verbose           Enable verbose output.
#   --start-idx <num>   Start index for participant selection.
#   --stop-idx <num>    Stop index for participant selection.

# This code is shared under the CC0 license
#
# Copyright (C) 2024, SIESTA workpackage 15 team

# This part of the script deals with possibly missing packages on-the-fly:
# It downloads them and puts them in a tempdir + adds the tempdir to the path
tdir <- tempdir()
.libPaths(c(tdir, .libPaths()))

hasoptparse  <- c("optparse") %in% rownames(installed.packages())
hasdplyr     <- c("dplyr") %in% rownames(installed.packages())
if (hasoptparse==FALSE)  {install.packages("optparse", lib=tdir, dependencies=TRUE, repos="https://cloud.r-project.org")}
if (hasdplyr==FALSE)     {install.packages("dplyr", lib=tdir, dependencies=TRUE, repos="https://cloud.r-project.org")}

# Load the required package for the option parsing
library("optparse", warn.conflicts = FALSE)
# Load the required package for column selection
library("dplyr", warn.conflicts = FALSE)

# Define the option parser
option_list <- list(
  make_option(c("-v", "--verbose"), action="store_true", default=FALSE,
              help="Print extra output"),
  make_option(c("--start-idx"), type="integer", default=0,
              help="Start index for participant selection", metavar="INTEGER"),
  make_option(c("--stop-idx"), type="integer", default=0,
              help="Stop index for participant selection", metavar="INTEGER")
)

# Parse the options
parser <- OptionParser(option_list=option_list, 
                       usage = "usage: %prog [options] input output level",
                       description = "This pipeline computes averages from the participants.tsv file.")
arguments <- parse_args(parser, positional_arguments = 3)
opts <- arguments$options
args <- arguments$args

# Check if help was requested
if (opts$help) {
  print_help(parser)
  quit(status = 0)
}

# Assign positional arguments
inputdir <- args[1]
outputdir <- args[2]
level <- args[3]

# Print verbose output if requested
if (opts$verbose) {
  cat("Verbose mode enabled\n")
  cat("Input directory:", inputdir, "\n")
  cat("Output directory:", outputdir, "\n")
  cat("Level:", level, "\n")
  cat("Starting index:", opts$'start-idx', "\n")
  cat("Stopping index:", opts$'stop-idx', "\n")
}

#
inputfile  <- file.path(inputdir, c("participants.tsv"))

# read table, deal with missing values
participants <- read.csv(inputfile, sep="\t", na.strings=c("n/a"))

# select the rows
if (opts$'stop-idx'>0) {
  participants <- participants[1:opts$'stop-idx', ]
}
if (opts$'start-idx'>0) {
  participants <- participants[opts$'start-idx':nrow(participants), ]
}

# print some of the columns
if (opts$verbose) {
  print(participants %>% select(1:5))
}

# create the output directory and its parents if they don't exist
dir.create(outputdir, recursive = TRUE, showWarnings = FALSE)

if (level == "participant") {
  print("nothing to do at the participant level, only creating participant-level output directories")
  for (i in 1:nrow(participants)) {
    dir.create(file.path(outputdir, participants$participant_id[i]), recursive = TRUE, showWarnings = FALSE)
  }

} else if (level == "group") {
  outputfile <- file.path(outputdir, "group", "results.tsv")
  dir.create(file.path(outputdir, "group"), recursive = TRUE, showWarnings = FALSE)

  # use the column names and capitalization from the original dataset
  # ignore missing values
  averagedage <- mean(participants$age, na.rm = TRUE)
  averagedHeight <- mean(participants$Height, na.rm = TRUE)
  averagedWeight <- mean(participants$Weight, na.rm = TRUE)

  # construct table with results
  result <- data.frame(averagedage, averagedHeight, averagedWeight)

  if (opts$verbose) {
    print(result)
  }

  # write the results to disk
  write.table(result, file=outputfile, sep="\t", col.names=FALSE, row.names=FALSE)
}
